pkg/containerruntimediscovery/containerddiscoverer.go

```diff
@@ -0,0 +1,191 @@
+	}, nil
+}
+
+func (cd *ContainerdDiscoverer) Images(ctx context.Context) ([]models.ContainerImageInfo, error) {
+	images, err := cd.client.ListImages(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list images: %w", err)
+	}
+
+	imageSet := map[string]models.ContainerImageInfo{}
+	for _, image := range images {
+		cii, err := cd.getContainerImageInfo(ctx, image)
+		if err != nil {
+			return nil, fmt.Errorf("unable to convert image %s to container image info: %w", image.Name(), err)
+		}
+
+		existing, ok := imageSet[cii.ImageID]
+		if ok {
+			merged, err := models.MergeContainerImage(existing, cii)
+			if err != nil {
+				return nil, fmt.Errorf("unable to merge image %v with %v: %w", existing, cii, err)
+			}
+			cii = merged
+		}
+		imageSet[cii.ImageID] = cii
+	}
+
+	result := []models.ContainerImageInfo{}
+	for _, image := range imageSet {
+		result = append(result, image)
+	}
+	return result, nil
+}
+
+func (cd *ContainerdDiscoverer) getContainerImageInfo(ctx context.Context, image containerd.Image) (models.ContainerImageInfo, error) {
+	configDescriptor, err := image.Config(ctx)
+	if err != nil {
+		return models.ContainerImageInfo{}, fmt.Errorf("failed to load image config descriptor: %w", err)
+	}
+	id := configDescriptor.Digest.String()
+
+	imageSpec, err := image.Spec(ctx)
+	if err != nil {
+		return models.ContainerImageInfo{}, fmt.Errorf("failed to load image spec: %w", err)
+	}
+
+	// NOTE(sambetts) We can not use image.Size as it gives us the size of
+	// the compressed layers and not the real size of the content.
+	snapshotter := cd.client.SnapshotService(containerd.DefaultSnapshotter)
+	size, err := imgutil.UnpackedImageSize(ctx, snapshotter, image)
+	if err != nil {
+		return models.ContainerImageInfo{}, fmt.Errorf("unable to determine size for image: %w", err)
+	}
+
+	repoTags, repoDigests := ParseImageReferences([]string{image.Name()})
+
+	return models.ContainerImageInfo{
+		ImageID:      id,
+		Architecture: utils.PointerTo(imageSpec.Architecture),
+		Labels:       convertTags(imageSpec.Config.Labels),
+		RepoTags:     &repoTags,
+		RepoDigests:  &repoDigests,
+		ObjectType:   "ContainerImageInfo",
+		Os:           utils.PointerTo(imageSpec.OS),
+		Size:         utils.PointerTo(int(size)),
+	}, nil
+}
+
+// ParseImageReferences parses a list of arbitrary image references and returns
+// the repotags and repodigests.
+func ParseImageReferences(refs []string) ([]string, []string) {
+	var tags, digests []string
+	for _, ref := range refs {
+		parsed, err := reference.ParseAnyReference(ref)
+		if err != nil {
+			continue
+		}
+		if _, ok := parsed.(reference.Canonical); ok {
+			digests = append(digests, parsed.String())
+		} else if _, ok := parsed.(reference.Tagged); ok {
+			tags = append(tags, parsed.String())
+		}
+	}
+	return tags, digests
+}
+
+func (cd *ContainerdDiscoverer) Containers(ctx context.Context) ([]models.ContainerInfo, error) {
+	containers, err := cd.client.Containers(ctx)
```

Comment:
I thought we talked about taking the containers info from the pods status

Reply:
PodStatus's don't give us all the information that we need without a lot of additional work, and actually not all the running containers are in the pod statuses either (for example the pause/sandbox container which makes up a pod is missing) therefore discovery through the container runtime is more accurate and efficient IMO